plugins/flatpak/gs-flatpak.c

```diff
@@ -2567,6 +2567,72 @@ gs_flatpak_related_should_download (GsFlatpak *self, GsApp *app, FlatpakRelatedR
 		}
 	}
 
+	/* EknServices apps should not be downloaded in an "Endless Platform"
+	 * update if no other installed apps use that branch of the runtime */
+	if (g_str_has_prefix (name, "com.endlessm.EknServices")) {
+		g_autoptr(GPtrArray) xrefs_installed = NULL;
+		g_autoptr(GError) error = NULL;
+		const gchar *services_name_ver;
+		gboolean found_app_that_uses_runtime = FALSE;
+
+		services_name_ver = name + 24;  /* strlen("com.endlessm.EknServices") */
+		if (*services_name_ver == '\0')
+			services_name_ver = "1";
+
+		xrefs_installed = flatpak_installation_list_installed_refs_by_kind (self->installation,
+		                                                                    FLATPAK_REF_KIND_APP,
+		                                                                    NULL, &error);
+		if (xrefs_installed == NULL) {
+			g_debug ("Couldn't get installed apps (%s), adding %s anyway",
+			         error->message, name);
+			return TRUE;
+		}
+
+		for (guint i = 0; i < xrefs_installed->len; i++) {
+			FlatpakInstalledRef *xref = g_ptr_array_index (xrefs_installed, i);
+			g_autoptr(GBytes) data = NULL;
+			g_autoptr(GsApp) runtime = NULL;
+			const gchar *str;
+			gsize len;
+
+			data = flatpak_installed_ref_load_metadata (xref, NULL, &error);
+			if (data == NULL) {
+				g_warning ("Couldn't get metadata for %s (%s), ",
+				           flatpak_ref_get_name (FLATPAK_REF (xref)),
+				           error->message);
+				g_clear_error (&error);
+				continue;
+			}
+
+			str = g_bytes_get_data (data, &len);
+			runtime = gs_flatpak_create_runtime_from_metadata (self,
+									   str, len,
+									   &error);
+			if (runtime == NULL) {
+				g_warning ("Couldn't get runtime for %s (%s)",
+				           flatpak_ref_get_name (FLATPAK_REF (xref)),
+				           error->message);
+				g_clear_error (&error);
+				continue;
+			}
+
+			/* Special case for EknServices 1 and eosplatform 3.1 */
+			if (g_strcmp0 (services_name_ver, "1") == 0 &&
+			    g_strcmp0 (gs_app_get_id (app), "com.endlessm.Platform") == 0 &&
+			    g_strcmp0 (gs_app_get_branch (app), "eos3.1")) {
+				found_app_that_uses_runtime = TRUE;
+				continue;
+			}
+
+			/* FIXME: can you compare pointer equality like this
+			 * because of the cache? */
+			if (runtime == app)
+				found_app_that_uses_runtime = TRUE;
+		}
+
+		return found_app_that_uses_runtime;
+	}
+
```

Comment:
I am afraid this will have an impact in performance: 1) Listing all installed refs with libflatpak is slow; 2) IO per each of those apps ditto.
I know this should happen only during installation/update, but still this code seems like a lot of complexity added "just" for skipping EknServices.

I understand that keeping a number of EknServices around may have an impact on the search, but I would rather include it's clean up as part of cleaning up unneeded legacy runtimes. So when we get rid of the eos3.2 Platform runtime, we should also uninstall EknServices (see `...get_list_for_remove`).

Reply:
OK, I agree with what you say and will change it.

However, the situation that I was trying to avoid was this:
- user doesn't have any SDK2 apps installed
- user updates Endless Platform
- com.endlessm.EknServices2 is in the list
- com.endlessm.EknServices2 gets installed as part of the Endless Platform update
- com.endlessm.EknServices2 pulls in SDK2 into the update

Do you know if that situation will actually happen, and/or what's a better way to avoid it?